prow/gerrit/reporter/reporter.go

```diff
@@ -62,33 +65,84 @@ func (c *Client) ShouldReport(pj *v1.ProwJob) bool {
-	return pj.ObjectMeta.Annotations["gerrit-id"] != "" &&
-		pj.ObjectMeta.Annotations["gerrit-instance"] != "" &&
-		pj.ObjectMeta.Labels["gerrit-revision"] != ""
+	if pj.ObjectMeta.Annotations["gerrit-id"] == "" ||
+		pj.ObjectMeta.Annotations["gerrit-instance"] == "" ||
+		pj.ObjectMeta.Labels["gerrit-revision"] == "" {
+		return false
+	}
+
+	// has other jobs on the same revision finished
+	//
+	selector := labels.Set{"gerrit-revision": pj.ObjectMeta.Labels["gerrit-revision"]}
+	pjs, err := c.lister.List(selector.AsSelector())
+	if err != nil {
+		logrus.WithError(err).Error("Cannot list prowjob with selector %v", selector)
+		return false
+	}
+
+	for _, pj := range pjs {
+		if pj.Status.State == v1.TriggeredState || pj.Status.State == v1.PendingState {
+			// not done yet
+			return false
+		}
+	}
+
+	return true
 }
 
 // Report will send the current prowjob status as a gerrit review
 func (c *Client) Report(pj *v1.ProwJob) error {
-	// TODO(krzyzacy): we should also do an aggregate report, as golang does in their repo
-	// see https://go-review.googlesource.com/c/go/+/132155
-	// also add ability to set code-review labels
-	// ref: https://github.com/kubernetes/test-infra/issues/9433
+	// If you are hitting here, which means the entire patchset has been finished :-)
 
-	if pj.Spec.Refs == nil {
-		return errors.New("no pj.Spec.Refs, not a presubmit job (should not happen?!)")
+	// list all prowjobs in the patchset
+	selector := labels.Set{"gerrit-revision": pj.ObjectMeta.Labels["gerrit-revision"]}
+	pjsOnRevision, err := c.lister.List(selector.AsSelector())
+	if err != nil {
+		logrus.WithError(err).Error("Cannot list prowjob with selector %v", selector)
+		return err
 	}
 
-	message := fmt.Sprintf("Job %s finished with %s\n Gubernator URL: %s", pj.Spec.Job, pj.Status.State, pj.Status.URL)
+	// generate an aggregated report:
+	total := len(pjsOnRevision)
+	success := 0
+	message := ""
+
+	for _, pjOnRevision := range pjsOnRevision {
+		if pjOnRevision.Status.PrevReportStates[c.GetName()] == pjOnRevision.Status.State {
+			logrus.Infof("Revision %s has been reported already", pj.ObjectMeta.Labels["gerrit-revision"])
+			return nil
+		}
+
+		if pjOnRevision.Status.State == v1.SuccessState {
+			success++
+		}
+
+		message = fmt.Sprintf("%s\nJob %s finished with %s -- URL: %s", message, pjOnRevision.Spec.Job, pjOnRevision.Status.State, pjOnRevision.Status.URL)
+	}
+
+	message = fmt.Sprintf("%d out of %d jobs passed!\n%s", success, total, message)
 
 	// report back
 	gerritID := pj.ObjectMeta.Annotations["gerrit-id"]
 	gerritInstance := pj.ObjectMeta.Annotations["gerrit-instance"]
 	gerritRevision := pj.ObjectMeta.Labels["gerrit-revision"]
 
-	logrus.Infof("Reporting job %s to instance %s on id %s with message %s", pj.Spec.Job, gerritInstance, gerritID, message)
-	if err := c.gc.SetReview(gerritInstance, gerritID, gerritRevision, message); err != nil {
-		logrus.Warnf("fail to set review")
-		return err
+	codeReview := "-1"
```

Comment:
I have a feeling this should be in a constant? or at least some comment to explain (or maybe we just assume all gerrit has this and the reader knows gerrit?) /shrug